File: Dart/Behavioral/iterator.dart
///Abstract class to define Iterator operations
abstract interface class Iterator {
  getNext<T>();
  bool hasMore();
}

/// Concrete number iterator implementation
class NumberIterator implements Iterator {
  int _index = 0;

  final List<int> numbersCollection;

  NumberIterator({required this.numbersCollection});

  /// Apply here concrete iteration rules
  @override
  getNext<T>() => numbersCollection[_index++];

  @override
  bool hasMore() => _index < numbersCollection.length;
}

/// Concrete number iterator implementation
class WordsIterator implements Iterator {
  int _index = 0;

  final List<String> wordsCollection;

  WordsIterator({required this.wordsCollection});

  /// Apply here concrete iteration rules
  @override
  getNext<T>() => wordsCollection[_index++];

  @override
  bool hasMore() => _index < wordsCollection.length;
}

/// Interface to define a collection.
/// Each collection defines an iterator.
abstract interface class IterableCollection {
  Iterator createIterator();
}

/// Concrete implementation of numbers collection
class NumbersCollection implements IterableCollection {
  NumbersCollection({required this.numbersCollection});

  final List<int> numbersCollection;

  /// Returns number iterator
  @override
  Iterator createIterator() =>
      NumberIterator(numbersCollection: numbersCollection);
}

/// Concrete implementation of words collection
class WordsCollection implements IterableCollection {
  WordsCollection({required this.wordsCollection});

  final List<String> wordsCollection;

  /// Returns word iterator
  @override
  Iterator createIterator() => WordsIterator(wordsCollection: wordsCollection);
}

void main() {
  final numbersCollection =
      NumbersCollection(numbersCollection: [1, 2, 3, 4, 5, 6, 7]);
  final numberIterator = numbersCollection.createIterator();

  final wordsCollection = WordsCollection(
      wordsCollection: ['one', 'two', 'three', 'four', 'five', 'six', 'seven']);
  final wordsIterator = wordsCollection.createIterator();

  print('-----Numbers Collection-----');
  while (numberIterator.hasMore()) {
    print(numberIterator.getNext());
  }

  print('-----Words Collection-----');
  while (wordsIterator.hasMore()) {
    print(wordsIterator.getNext());
  }

  /// Here is the result:
/*
-----Numbers Collection-----
1
2
3
4
5
6
7
-----Words Collection-----
one
two
three
four
five
six
seven
*/
}

File: Swift/Behavioral/iterator.swift
//
//  iterator.swift
//
//

// Protocol to define Iterator operations
protocol Iterator {
    associatedtype Element
    
    mutating func getNext() -> Element
    
    func hasMore() -> Bool
}

// Concrete number iterator implementation
struct NumberIterator: Iterator {
    private var index = 0

    let numbersCollection: [Int]

    init(numbersCollection: [Int]) {
        self.numbersCollection = numbersCollection
    }

    // Apply here concrete iteration rules
    mutating func getNext() -> Int {
        let value = numbersCollection[index]
        
        index += 1

        return value
    }

    func hasMore() -> Bool { index < numbersCollection.count }
}

// Concrete words iterator implementation
struct WordsIterator: Iterator {
    private var index = 0

    let wordsCollection: [String]

    init(wordsCollection: [String]) {
        self.wordsCollection = wordsCollection
    }

    // Apply here concrete iteration rules
    mutating func getNext() -> String {
        let value = wordsCollection[index]
        
        index += 1
        
        return value
    }

    func hasMore() -> Bool { index < wordsCollection.count }
}

// Protocol to define a collection.
// Each collection defines an iterator.
protocol IterableCollection {
    associatedtype IteratorType: Iterator
    func createIterator() -> IteratorType
}

// Concrete implementation of numbers collection
struct NumbersCollection: IterableCollection {
    let numbersCollection: [Int]

    init(numbersCollection: [Int]) {
        self.numbersCollection = numbersCollection
    }

    // Returns number iterator
    func createIterator() -> NumberIterator {
        NumberIterator(numbersCollection: numbersCollection)
    }
}

// Concrete implementation of words collection
struct WordsCollection: IterableCollection {
    let wordsCollection: [String]

    init(wordsCollection: [String]) {
        self.wordsCollection = wordsCollection
    }

    // Returns word iterator
    func createIterator() -> WordsIterator {
        WordsIterator(wordsCollection: wordsCollection)
    }
}

// Demo

let numbersCollection = NumbersCollection(numbersCollection: [1, 2, 3, 4, 5, 6, 7])
var numberIterator = numbersCollection.createIterator()

let wordsCollection = WordsCollection(
    wordsCollection: ["one", "two", "three", "four", "five", "six", "seven"])
var wordsIterator = wordsCollection.createIterator()

print("-----Numbers Collection-----")
while numberIterator.hasMore() {
    print(numberIterator.getNext())
}

print("-----Words Collection-----")
while wordsIterator.hasMore() {
    print(wordsIterator.getNext())
}

// Here is the result:

// -----Numbers Collection-----
// 1
// 2
// 3
// 4
// 5
// 6
// 7
// -----Words Collection-----
// one
// two
// three
// four
// five
// six
// seven
